{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "현재 상태:\n",
    "\n",
    "Tumor 디렉토리:\n",
    "\n",
    "    - SRR30510137: ✅ 완료\n",
    "\n",
    "    - SRR30510151: ✅ 완료\n",
    "\n",
    "    - SRR30510181: ✅ 완료\n",
    "\n",
    "Normal 디렉토리:\n",
    "\n",
    "    - SRR28602760: ✅ 완료\n",
    "\n",
    "    - SRR28602761: ✅ 완료\n",
    "\n",
    "    - SRR28602767: ✅ 완료\n",
    "\n",
    "각 디렉토리 별로 아래 단계가 완료된 상태입니다.\n",
    "\n",
    "- fastqc (품질 체크)\n",
    "\n",
    "- trim_galore (트리밍)\n",
    "\n",
    "- STAR (정렬 및 BAM 파일 생성)\n",
    "\n",
    "- featureCounts (유전자 발현량 정량화, counts.txt 생성)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample           SRR30510137  SRR30510151  SRR30510181  SRR28602760  \\\n",
      "Geneid                                                                \n",
      "ENSG00000279928            6            0            2            2   \n",
      "ENSG00000228037            0            0            7           15   \n",
      "ENSG00000142611         2403          735          168         1128   \n",
      "ENSG00000284616            0            0            0            0   \n",
      "ENSG00000157911         1369          537          889          805   \n",
      "\n",
      "Sample           SRR28602761  SRR28602767  \n",
      "Geneid                                     \n",
      "ENSG00000279928            2            6  \n",
      "ENSG00000228037            8            6  \n",
      "ENSG00000142611         1247          504  \n",
      "ENSG00000284616            0            0  \n",
      "ENSG00000157911         1177         1301  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "\n",
    "tumor_paths = glob.glob('./Tumor/*/counts.txt')\n",
    "normal_paths = glob.glob('./Normal/*/counts.txt')\n",
    "\n",
    "usecols = ['Geneid', 'SRR']\n",
    "\n",
    "counts_tumor = []\n",
    "for path in tumor_paths:\n",
    "    sample_name = os.path.basename(os.path.dirname(path))\n",
    "    df = pd.read_csv(path, sep='\\t', comment='#', usecols=[0, 6], names=['Geneid', sample_name], header=0)\n",
    "    counts_tumor.append(df.set_index('Geneid'))\n",
    "\n",
    "counts_tumor_df = pd.concat(counts_tumor, axis=1)\n",
    "\n",
    "counts_normal = []\n",
    "for path in normal_paths:\n",
    "    sample_name = os.path.basename(os.path.dirname(path))\n",
    "    df = pd.read_csv(path, sep='\\t', comment='#', usecols=[0, 6], names=['Geneid', sample_name], header=0)\n",
    "    counts_normal.append(df.set_index('Geneid'))\n",
    "\n",
    "counts_normal_df = pd.concat(counts_normal, axis=1)\n",
    "\n",
    "combined_counts_df = pd.concat([counts_tumor_df, counts_normal_df], axis=1)\n",
    "combined_counts_df.columns.name = 'Sample'\n",
    "\n",
    "print(combined_counts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_counts_df.columns = [\n",
    "    'Tumor_SRR30510137',\n",
    "    'Tumor_SRR30510151',\n",
    "    'Tumor_SRR30510181',\n",
    "    'Normal_SRR28602760',\n",
    "    'Normal_SRR28602761',\n",
    "    'Normal_SRR28602767'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tumor_SRR30510137</th>\n",
       "      <th>Tumor_SRR30510151</th>\n",
       "      <th>Tumor_SRR30510181</th>\n",
       "      <th>Normal_SRR28602760</th>\n",
       "      <th>Normal_SRR28602761</th>\n",
       "      <th>Normal_SRR28602767</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000279928</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228037</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142611</th>\n",
       "      <td>2403</td>\n",
       "      <td>735</td>\n",
       "      <td>168</td>\n",
       "      <td>1128</td>\n",
       "      <td>1247</td>\n",
       "      <td>504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284616</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000157911</th>\n",
       "      <td>1369</td>\n",
       "      <td>537</td>\n",
       "      <td>889</td>\n",
       "      <td>805</td>\n",
       "      <td>1177</td>\n",
       "      <td>1301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Tumor_SRR30510137  Tumor_SRR30510151  Tumor_SRR30510181  \\\n",
       "Geneid                                                                     \n",
       "ENSG00000279928                  6                  0                  2   \n",
       "ENSG00000228037                  0                  0                  7   \n",
       "ENSG00000142611               2403                735                168   \n",
       "ENSG00000284616                  0                  0                  0   \n",
       "ENSG00000157911               1369                537                889   \n",
       "\n",
       "                 Normal_SRR28602760  Normal_SRR28602761  Normal_SRR28602767  \n",
       "Geneid                                                                       \n",
       "ENSG00000279928                   2                   2                   6  \n",
       "ENSG00000228037                  15                   8                   6  \n",
       "ENSG00000142611                1128                1247                 504  \n",
       "ENSG00000284616                   0                   0                   0  \n",
       "ENSG00000157911                 805                1177                1301  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_counts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found existing installation: gprofiler 1.2.2\n",
      "Uninstalling gprofiler-1.2.2:\n",
      "  Would remove:\n",
      "    /home/min/anaconda3/envs/BIO/lib/python3.10/site-packages/gprofiler-1.2.2.dist-info/*\n",
      "    /home/min/anaconda3/envs/BIO/lib/python3.10/site-packages/gprofiler/*\n",
      "Proceed (Y/n)? \u001b[31mERROR: Operation cancelled by user\u001b[0m\u001b[31m\n",
      "\u001b[0m^C\n",
      "Collecting gprofiler-official\n",
      "  Using cached gprofiler_official-1.0.0-py3-none-any.whl.metadata (11 kB)\n",
      "Requirement already satisfied: requests in /home/min/anaconda3/envs/BIO/lib/python3.10/site-packages (from gprofiler-official) (2.32.3)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/min/anaconda3/envs/BIO/lib/python3.10/site-packages (from requests->gprofiler-official) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/min/anaconda3/envs/BIO/lib/python3.10/site-packages (from requests->gprofiler-official) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /home/min/anaconda3/envs/BIO/lib/python3.10/site-packages (from requests->gprofiler-official) (2.3.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/min/anaconda3/envs/BIO/lib/python3.10/site-packages (from requests->gprofiler-official) (2025.1.31)\n",
      "Using cached gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
      "Installing collected packages: gprofiler-official\n",
      "Successfully installed gprofiler-official-1.0.0\n"
     ]
    }
   ],
   "source": [
    "!pip uninstall gprofiler\n",
    "!pip install gprofiler-official"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential Expression Analysis (발현차이 분석)\n",
    "\n",
    "- DESeq2, edgeR, limma-voom이 대표적입니다.\n",
    "\n",
    "- 그룹간(종양 vs 정상) 유의하게 발현이 다른 유전자를 찾음\n",
    "\n",
    "- Fold-change 및 adjusted p-value가 중요합니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: MatrixGenerics\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 이미 설치된 상태에서 실행 가능한 코드\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "pandas2ri.activate()\n",
    "\n",
    "r = robjects.r\n",
    "r('''\n",
    "library(DESeq2)\n",
    "run_DESeq2 <- function(countData, condition) {\n",
    "  colData <- data.frame(row.names=colnames(countData), condition=condition)\n",
    "  dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~ condition)\n",
    "  dds <- DESeq(dds)\n",
    "  res <- results(dds)\n",
    "  return(as.data.frame(res))\n",
    "}\n",
    "''')\n",
    "\n",
    "run_DESeq2 = r['run_DESeq2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "cannot import name 'GProfiler' from 'gprofiler' (/home/min/anaconda3/envs/BIO/lib/python3.10/site-packages/gprofiler/__init__.py)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mgprofiler\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m GProfiler\n\u001b[1;32m      2\u001b[0m gp \u001b[38;5;241m=\u001b[39m GProfiler(return_dataframe\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      3\u001b[0m results \u001b[38;5;241m=\u001b[39m gp\u001b[38;5;241m.\u001b[39mprofile(organism\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhsapiens\u001b[39m\u001b[38;5;124m'\u001b[39m, query\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mBRCA1\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTP53\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEGFR\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "\u001b[0;31mImportError\u001b[0m: cannot import name 'GProfiler' from 'gprofiler' (/home/min/anaconda3/envs/BIO/lib/python3.10/site-packages/gprofiler/__init__.py)"
     ]
    }
   ],
   "source": [
    "from gprofiler import GProfiler\n",
    "gp = GProfiler(return_dataframe=True)\n",
    "results = gp.profile(organism='hsapiens', query=['BRCA1', 'TP53', 'EGFR'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BIO",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
