{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "현재 상태:\n",
    "\n",
    "Tumor 디렉토리:\n",
    "\n",
    "    - SRR30510137: ✅ 완료\n",
    "\n",
    "    - SRR30510151: ✅ 완료\n",
    "\n",
    "    - SRR30510181: ✅ 완료\n",
    "\n",
    "Normal 디렉토리:\n",
    "\n",
    "    - SRR28602760: ✅ 완료\n",
    "\n",
    "    - SRR28602761: ✅ 완료\n",
    "\n",
    "    - SRR28602767: ✅ 완료\n",
    "\n",
    "각 디렉토리 별로 아래 단계가 완료된 상태입니다.\n",
    "\n",
    "- fastqc (품질 체크)\n",
    "\n",
    "- trim_galore (트리밍)\n",
    "\n",
    "- STAR (정렬 및 BAM 파일 생성)\n",
    "\n",
    "- featureCounts (유전자 발현량 정량화, counts.txt 생성)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample           SRR30510137  SRR30510151  SRR30510181  SRR28602760  \\\n",
      "Geneid                                                                \n",
      "ENSG00000279928            6            0            2            2   \n",
      "ENSG00000228037            0            0            7           15   \n",
      "ENSG00000142611         2403          735          168         1128   \n",
      "ENSG00000284616            0            0            0            0   \n",
      "ENSG00000157911         1369          537          889          805   \n",
      "\n",
      "Sample           SRR28602761  SRR28602767  \n",
      "Geneid                                     \n",
      "ENSG00000279928            2            6  \n",
      "ENSG00000228037            8            6  \n",
      "ENSG00000142611         1247          504  \n",
      "ENSG00000284616            0            0  \n",
      "ENSG00000157911         1177         1301  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "\n",
    "tumor_paths = glob.glob('./Tumor/*/counts.txt')\n",
    "normal_paths = glob.glob('./Normal/*/counts.txt')\n",
    "\n",
    "usecols = ['Geneid', 'SRR']\n",
    "\n",
    "counts_tumor = []\n",
    "for path in tumor_paths:\n",
    "    sample_name = os.path.basename(os.path.dirname(path))\n",
    "    df = pd.read_csv(path, sep='\\t', comment='#', usecols=[0, 6], names=['Geneid', sample_name], header=0)\n",
    "    counts_tumor.append(df.set_index('Geneid'))\n",
    "\n",
    "counts_tumor_df = pd.concat(counts_tumor, axis=1)\n",
    "\n",
    "counts_normal = []\n",
    "for path in normal_paths:\n",
    "    sample_name = os.path.basename(os.path.dirname(path))\n",
    "    df = pd.read_csv(path, sep='\\t', comment='#', usecols=[0, 6], names=['Geneid', sample_name], header=0)\n",
    "    counts_normal.append(df.set_index('Geneid'))\n",
    "\n",
    "counts_normal_df = pd.concat(counts_normal, axis=1)\n",
    "\n",
    "combined_counts_df = pd.concat([counts_tumor_df, counts_normal_df], axis=1)\n",
    "combined_counts_df.columns.name = 'Sample'\n",
    "\n",
    "print(combined_counts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_counts_df.columns = [\n",
    "    'Tumor_SRR30510137',\n",
    "    'Tumor_SRR30510151',\n",
    "    'Tumor_SRR30510181',\n",
    "    'Normal_SRR28602760',\n",
    "    'Normal_SRR28602761',\n",
    "    'Normal_SRR28602767'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Geneid가 인덱스인 상태에서 바로 저장\n",
    "combined_counts_df.to_csv(\"Result/combined_counts_df.csv\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tumor_SRR30510137</th>\n",
       "      <th>Tumor_SRR30510151</th>\n",
       "      <th>Tumor_SRR30510181</th>\n",
       "      <th>Normal_SRR28602760</th>\n",
       "      <th>Normal_SRR28602761</th>\n",
       "      <th>Normal_SRR28602767</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000279928</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228037</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142611</th>\n",
       "      <td>2403</td>\n",
       "      <td>735</td>\n",
       "      <td>168</td>\n",
       "      <td>1128</td>\n",
       "      <td>1247</td>\n",
       "      <td>504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284616</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000157911</th>\n",
       "      <td>1369</td>\n",
       "      <td>537</td>\n",
       "      <td>889</td>\n",
       "      <td>805</td>\n",
       "      <td>1177</td>\n",
       "      <td>1301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>1719</td>\n",
       "      <td>790</td>\n",
       "      <td>351</td>\n",
       "      <td>271</td>\n",
       "      <td>619</td>\n",
       "      <td>1450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275987</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268674</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277475</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275405</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>96</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62754 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Tumor_SRR30510137  Tumor_SRR30510151  Tumor_SRR30510181  \\\n",
       "Geneid                                                                     \n",
       "ENSG00000279928                  6                  0                  2   \n",
       "ENSG00000228037                  0                  0                  7   \n",
       "ENSG00000142611               2403                735                168   \n",
       "ENSG00000284616                  0                  0                  0   \n",
       "ENSG00000157911               1369                537                889   \n",
       "...                            ...                ...                ...   \n",
       "ENSG00000271254               1719                790                351   \n",
       "ENSG00000275987                  0                  0                  0   \n",
       "ENSG00000268674                  0                  0                  0   \n",
       "ENSG00000277475                  4                  0                  0   \n",
       "ENSG00000275405                  8                  1                 96   \n",
       "\n",
       "                 Normal_SRR28602760  Normal_SRR28602761  Normal_SRR28602767  \n",
       "Geneid                                                                       \n",
       "ENSG00000279928                   2                   2                   6  \n",
       "ENSG00000228037                  15                   8                   6  \n",
       "ENSG00000142611                1128                1247                 504  \n",
       "ENSG00000284616                   0                   0                   0  \n",
       "ENSG00000157911                 805                1177                1301  \n",
       "...                             ...                 ...                 ...  \n",
       "ENSG00000271254                 271                 619                1450  \n",
       "ENSG00000275987                   2                   4                   0  \n",
       "ENSG00000268674                   0                   0                   0  \n",
       "ENSG00000277475                   1                   0                   8  \n",
       "ENSG00000275405                   0                   0                   0  \n",
       "\n",
       "[62754 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_counts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### combined_counts_df 데이터에서 중복 Geneid 제거하기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Tumor_SRR30510137</th>\n",
       "      <th>Tumor_SRR30510151</th>\n",
       "      <th>Tumor_SRR30510181</th>\n",
       "      <th>Normal_SRR28602760</th>\n",
       "      <th>Normal_SRR28602761</th>\n",
       "      <th>Normal_SRR28602767</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000228037</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>2403</td>\n",
       "      <td>735</td>\n",
       "      <td>168</td>\n",
       "      <td>1128</td>\n",
       "      <td>1247</td>\n",
       "      <td>504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000284616</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>1369</td>\n",
       "      <td>537</td>\n",
       "      <td>889</td>\n",
       "      <td>805</td>\n",
       "      <td>1177</td>\n",
       "      <td>1301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62749</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>1719</td>\n",
       "      <td>790</td>\n",
       "      <td>351</td>\n",
       "      <td>271</td>\n",
       "      <td>619</td>\n",
       "      <td>1450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62750</th>\n",
       "      <td>ENSG00000275987</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62751</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62752</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62753</th>\n",
       "      <td>ENSG00000275405</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>96</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62754 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Geneid  Tumor_SRR30510137  Tumor_SRR30510151  \\\n",
       "0      ENSG00000279928                  6                  0   \n",
       "1      ENSG00000228037                  0                  0   \n",
       "2      ENSG00000142611               2403                735   \n",
       "3      ENSG00000284616                  0                  0   \n",
       "4      ENSG00000157911               1369                537   \n",
       "...                ...                ...                ...   \n",
       "62749  ENSG00000271254               1719                790   \n",
       "62750  ENSG00000275987                  0                  0   \n",
       "62751  ENSG00000268674                  0                  0   \n",
       "62752  ENSG00000277475                  4                  0   \n",
       "62753  ENSG00000275405                  8                  1   \n",
       "\n",
       "       Tumor_SRR30510181  Normal_SRR28602760  Normal_SRR28602761  \\\n",
       "0                      2                   2                   2   \n",
       "1                      7                  15                   8   \n",
       "2                    168                1128                1247   \n",
       "3                      0                   0                   0   \n",
       "4                    889                 805                1177   \n",
       "...                  ...                 ...                 ...   \n",
       "62749                351                 271                 619   \n",
       "62750                  0                   2                   4   \n",
       "62751                  0                   0                   0   \n",
       "62752                  0                   1                   0   \n",
       "62753                 96                   0                   0   \n",
       "\n",
       "       Normal_SRR28602767  \n",
       "0                       6  \n",
       "1                       6  \n",
       "2                     504  \n",
       "3                       0  \n",
       "4                    1301  \n",
       "...                   ...  \n",
       "62749                1450  \n",
       "62750                   0  \n",
       "62751                   0  \n",
       "62752                   8  \n",
       "62753                   0  \n",
       "\n",
       "[62754 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_counts_df = pd.read_csv(\"Result/combined_counts_df.csv\")\n",
    "\n",
    "combined_counts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 정규화(Normalization)\n",
    "\n",
    "- RNA-Seq 데이터 분석의 필수 단계입니다. \n",
    "- 다양한 방법 중 CPM(Counts Per Million)과 log 변환을 추천합니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 Tumor_SRR30510137  Tumor_SRR30510151  Tumor_SRR30510181  \\\n",
      "Geneid                                                                     \n",
      "ENSG00000279928           0.114254           0.000000           0.083805   \n",
      "ENSG00000228037           0.000000           0.000000           0.274213   \n",
      "ENSG00000142611           5.087772           3.979975           2.590725   \n",
      "ENSG00000284616           0.000000           0.000000           0.000000   \n",
      "ENSG00000157911           4.307749           3.560476           4.785833   \n",
      "\n",
      "                 Normal_SRR28602760  Normal_SRR28602761  Normal_SRR28602767  \n",
      "Geneid                                                                       \n",
      "ENSG00000279928            0.038284            0.031208            0.109349  \n",
      "ENSG00000228037            0.265062            0.120976            0.109349  \n",
      "ENSG00000142611            4.014973            3.871297            2.928702  \n",
      "ENSG00000284616            0.000000            0.000000            0.000000  \n",
      "ENSG00000157911            3.563635            3.793801            4.175820  \n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "combined_counts_df = pd.read_csv('Result/combined_counts_df.csv', index_col=0)\n",
    "\n",
    "def cpm(counts):\n",
    "    return (counts / counts.sum()) * 1e6\n",
    "\n",
    "normalized_df = combined_counts_df.apply(cpm, axis=0)\n",
    "normalized_log_df = np.log2(normalized_df + 1)\n",
    "\n",
    "normalized_log_df.to_csv('Result/normalized_counts.csv', index=False)\n",
    "\n",
    "print(normalized_log_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA(Principal Component Analysis)\n",
    "\n",
    "정규화 후 PCA를 통해 샘플 간 유사성을 탐색하고, 각 군(Tumor vs Normal)의 명확한 분리를 확인할 수 있음"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'normalized_counts.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 5\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mmatplotlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpyplot\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mplt\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mseaborn\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01msns\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m data \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mnormalized_counts.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m pca \u001b[38;5;241m=\u001b[39m PCA(n_components\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m)\n\u001b[1;32m      8\u001b[0m pca_result \u001b[38;5;241m=\u001b[39m pca\u001b[38;5;241m.\u001b[39mfit_transform(data\u001b[38;5;241m.\u001b[39mT)\n",
      "File \u001b[0;32m~/anaconda3/envs/BIO/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1013\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m   1014\u001b[0m     dialect,\n\u001b[1;32m   1015\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1022\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m   1023\u001b[0m )\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/BIO/lib/python3.10/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    619\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    623\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/anaconda3/envs/BIO/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/BIO/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1880\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1878\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1879\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1880\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1882\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1883\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1884\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1885\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1886\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1887\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1888\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1889\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1890\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1891\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/anaconda3/envs/BIO/lib/python3.10/site-packages/pandas/io/common.py:873\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    868\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    869\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    870\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    871\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    872\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 873\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    874\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    875\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    876\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    877\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    878\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    879\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    880\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    881\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    882\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'normalized_counts.csv'"
     ]
    }
   ],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "data = pd.read_csv('normalized_counts.csv')\n",
    "\n",
    "pca = PCA(n_components=2)\n",
    "pca_result = pca.fit_transform(data.T)\n",
    "\n",
    "pca_df = pd.DataFrame(pca_result, columns=['PC1', 'PC2'])\n",
    "pca_df['Sample'] = ['Tumor', 'Tumor', 'Tumor', 'Normal', 'Normal', 'Normal']\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "sns.scatterplot(data=pca_df, x='PC1', y='PC2', hue='Sample', s=100)\n",
    "\n",
    "plt.title('PCA of Glioblastoma RNA-seq Samples')\n",
    "plt.xlabel('Principal Component 1')\n",
    "plt.ylabel('Principal Component 2')\n",
    "plt.legend(title='Sample type')\n",
    "plt.grid(True)\n",
    "\n",
    "plt.savefig('PCA_plot.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DESeq2 (R)\n",
    "\n",
    "DESeq2는 RNA 시퀀싱 데이터를 분석할 때 사용하는 대표적인 R 패키지입니다. \n",
    "\n",
    "특히, **2개 이상의 그룹 간 유전자 발현 수준의 차이를 통게적으로 평가할 때 사용함.**\n",
    "\n",
    "예를 들어, 질병 조직과 정상 조직 사이에서 어떤 유전자들이 통계적으로 의미 있게 차이가 나는지를 파악할 때 사용합니다.\n",
    "\n",
    "이런 유전자들을 **차등 발현 유전자(Differentially Expressed Genes, DEGs)** 라고 부릅니다.\n",
    "\n",
    "---\n",
    "\n",
    "#### DESeq2를 사용하는 시점과 목적 \n",
    "\n",
    "- RNA-seq 데이터를 분석하여 질병과 정상 조직 간의 유전자 차이 파악이 필요할 때 사용합니다. \n",
    "\n",
    "- 약물 처리군과 대조군 간의 유전자 발현 변화를 평가할 때 사용합니다. \n",
    "\n",
    "- 돌연변이 유무, 유전자 Knock-down, 유전자 Over-expression 등 실험 조건 간 유전자 발현 수준 비교 시 사용합니다. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000279928</th>\n",
       "      <td>2.758013</td>\n",
       "      <td>-0.405592</td>\n",
       "      <td>1.749998</td>\n",
       "      <td>-0.231767</td>\n",
       "      <td>0.816719</td>\n",
       "      <td>0.894558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228037</th>\n",
       "      <td>6.388304</td>\n",
       "      <td>-1.640095</td>\n",
       "      <td>1.805583</td>\n",
       "      <td>-0.908346</td>\n",
       "      <td>0.363695</td>\n",
       "      <td>0.533550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142611</th>\n",
       "      <td>921.031206</td>\n",
       "      <td>-0.018921</td>\n",
       "      <td>0.863661</td>\n",
       "      <td>-0.021908</td>\n",
       "      <td>0.982522</td>\n",
       "      <td>0.990358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284616</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000157911</th>\n",
       "      <td>981.995279</td>\n",
       "      <td>-0.165080</td>\n",
       "      <td>0.582484</td>\n",
       "      <td>-0.283407</td>\n",
       "      <td>0.776864</td>\n",
       "      <td>0.869298</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "ENSG00000279928    2.758013       -0.405592  1.749998 -0.231767  0.816719   \n",
       "ENSG00000228037    6.388304       -1.640095  1.805583 -0.908346  0.363695   \n",
       "ENSG00000142611  921.031206       -0.018921  0.863661 -0.021908  0.982522   \n",
       "ENSG00000284616    0.000000             NaN       NaN       NaN       NaN   \n",
       "ENSG00000157911  981.995279       -0.165080  0.582484 -0.283407  0.776864   \n",
       "\n",
       "                     padj  \n",
       "ENSG00000279928  0.894558  \n",
       "ENSG00000228037  0.533550  \n",
       "ENSG00000142611  0.990358  \n",
       "ENSG00000284616       NaN  \n",
       "ENSG00000157911  0.869298  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "res_df = pd.read_csv('Result/DESeq2_results.csv', index_col=0)\n",
    "res_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R에서 DESeq2로 분석한 결과를 파이썬으로 불러오기 \n",
    "\n",
    "- 유의한 DEGs를 필터링 \n",
    "\n",
    "- 필터링을 통해 의미있는 유전자 집합을 추출하고, GO enrichment 또는 Pathway 분석 등의 추가 분석으로 나아갈 수 있습니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "유의한 차등 발현 유전자 수: 8231\n",
      "종양에서 증가된 유전자 수: 3287\n",
      "종양에서 감소된 유전자 수: 4944\n"
     ]
    }
   ],
   "source": [
    "# 유의한 유전자만 필터링 (padj < 0.05 & 절대값 log2FoldChange > 1)\n",
    "sig_genes = res_df[(res_df['padj'] < 0.05) & (abs(res_df['log2FoldChange']) > 1)]\n",
    "\n",
    "# 종양에서 높게 발현된 유전자\n",
    "upregulated = sig_genes[sig_genes['log2FoldChange'] > 0]\n",
    "\n",
    "# 종양에서 낮게 발현된 유전자\n",
    "downregulated = sig_genes[sig_genes['log2FoldChange'] < 0]\n",
    "\n",
    "print(\"유의한 차등 발현 유전자 수:\", len(sig_genes))\n",
    "print(\"종양에서 증가된 유전자 수:\", len(upregulated))\n",
    "print(\"종양에서 감소된 유전자 수:\", len(downregulated))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 기능적 해석 (Functional Enrichment Analysis)\n",
    "\n",
    "- 유의한 유전자들이 어떤 생물학적 과정(Gene Ontology, GO) 또는 대사 경로(Pathway)와 연관되어 있는지를 파악함.\n",
    "\n",
    "#### GProfiler  \n",
    "\n",
    "- \"유전자 목록을 넣으면, 이 유전자들이 공통적으로 어떤 기능에 관여하는지\"를 분석하는 프로그램입니다.\n",
    "\n",
    "- 현재 Glioblastoma RNA-seq 분석에서 얻은 유의한 유전자 목록을 입력하고, 이 유전자들이 공통적으로 어떤 기능에 연관되어 있는지를 분석하려는 단계입니다.\n",
    "\n",
    "- `sources` 옵션에는 분석 방법이나 알고리즘을 넣는 것이 아니라, 어떤 생물학적 데이터베이스를 이용해서 유전자 기능을 해석할지 지정하는 것입니다. \n",
    "\n",
    "- `query`는 유전자 목록을 넣고 `sources`는 분석에 사용할 데이터베이스를 넣음."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  source      native                              name       p_value  \\\n",
      "0  GO:BP  GO:0065007             biological regulation  1.059754e-50   \n",
      "1  GO:BP  GO:0050789  regulation of biological process  7.477963e-47   \n",
      "2  GO:BP  GO:0032502             developmental process  2.207014e-46   \n",
      "3  GO:BP  GO:0048856  anatomical structure development  3.787809e-46   \n",
      "4  GO:BP  GO:0050794    regulation of cellular process  1.424319e-43   \n",
      "\n",
      "   significant                                        description  term_size  \\\n",
      "0         True  \"Any process that modulates a measurable attri...      12671   \n",
      "1         True  \"Any process that modulates the frequency, rat...      12278   \n",
      "2         True  \"A biological process whose specific outcome i...       6478   \n",
      "3         True  \"The biological process whose specific outcome...       5924   \n",
      "4         True  \"Any process that modulates the frequency, rat...      11876   \n",
      "\n",
      "   query_size  intersection_size  effective_domain_size  precision    recall  \\\n",
      "0        4718               3296                  21017   0.698601  0.260122   \n",
      "1        4718               3195                  21017   0.677194  0.260222   \n",
      "2        4718               1874                  21017   0.397202  0.289287   \n",
      "3        4718               1739                  21017   0.368588  0.293552   \n",
      "4        4718               3093                  21017   0.655574  0.260441   \n",
      "\n",
      "     query                   parents  \n",
      "0  query_1              [GO:0008150]  \n",
      "1  query_1  [GO:0008150, GO:0065007]  \n",
      "2  query_1              [GO:0008150]  \n",
      "3  query_1              [GO:0032502]  \n",
      "4  query_1  [GO:0009987, GO:0050789]  \n"
     ]
    }
   ],
   "source": [
    "from gprofiler import GProfiler\n",
    "\n",
    "significant_genes = sig_genes.index.tolist()\n",
    "\n",
    "gp = GProfiler(return_dataframe=True)\n",
    "enrichment_results = gp.profile(\n",
    "    organism='hsapiens',\n",
    "    query=significant_genes,\n",
    "    sources=['GO:BP', 'KEGG', 'REAC']\n",
    ")\n",
    "\n",
    "print(enrichment_results.head())\n",
    "enrichment_results.to_csv(\"enrichment_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BIO",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
